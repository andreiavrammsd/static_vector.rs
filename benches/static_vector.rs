use criterion::{Criterion, criterion_group, criterion_main};
use static_vector::Vec;

fn bench_static_vector(c: &mut Criterion) {
    let mut vec = Vec::<i32, 10>::new();

    c.bench_function("push and clear", |b| {
        b.iter(|| {
            vec.push(&1).unwrap();
            vec.push(&2).unwrap();
            vec.push(&3).unwrap();
            vec.push(&4).unwrap();
            vec.push(&5).unwrap();
            vec.push(&6).unwrap();
            vec.clear();
        })
    });
}

criterion_group!(benches, bench_static_vector);
criterion_main!(benches);
