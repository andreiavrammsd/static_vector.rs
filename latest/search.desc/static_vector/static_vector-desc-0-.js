searchState.loadedDescShard("static_vector", 0, "<code>static_vector</code>\nAttempted to push to a full vector\nImmutable iterator over a <code>Vec</code>.\nMutable iterator over a <code>Vec</code>.\nAttempted to resize the vector to a length greater than …\nA stack-allocated vector with fixed capacity and dynamic …\nReturns the maximum number of elements the vector can …\nRemoves all elements. Size will be zero.\nCreates an empty <code>Vec</code>. Equivalent to <code>Vec::new()</code>.\nReturns a reference to the first element in the vector, or …\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns a reference to the element at the specified <code>index</code>, …\nReturns a mutable reference to the element at the …\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nReturns whether the vector has no elements or any.\nReturns whether the vector is at maximum capacity.\nReturns an iterator over immutable references to the …\nReturns an iterator over mutable references to the …\nReturns a reference to the last element in the vector, or …\nReturns the maximum number of elements the vector currenly …\nCreates a new empty <code>Vec</code> with maximum <code>CAPACITY</code> elements of …\nCreates immutable iterator.\nCreates mutable iterator.\nReturns (and removes) the last element from the vector, or …\nReturns (and removes) the last element from the vector if …\nAdds a clone of the given <code>value</code> to the end of the vector.\nResizes the vector to the <code>new_length</code>.")