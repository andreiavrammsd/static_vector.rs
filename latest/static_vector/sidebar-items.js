window.SIDEBAR_ITEMS = {"struct":["CapacityExceededError","Iter","IterMut","LengthTooLargeError","Vec"]};