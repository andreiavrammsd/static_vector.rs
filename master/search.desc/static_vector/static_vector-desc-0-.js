searchState.loadedDescShard("static_vector", 0, "<code>static_vector</code>\nError for when the vector is full or the requested …\nImmutable iterator over a <code>Vec</code>.\nMutable iterator over a <code>Vec</code>.\nA stack-allocated vector with fixed capacity and dynamic …\nReturns a mutable slice of the entire vector.\nReturns a slice of the entire vector.\nReturns the maximum number of elements the vector can …\nRemoves all elements. Size will be zero.\nCreates an empty <code>Vec</code>. Equivalent to <code>Vec::new()</code>.\nInserts elements of given slice at the end of the vector.\nReturns a reference to the first element in the vector, or …\nReturns a mutable reference to the first element in the …\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns a reference to the element at the specified <code>index</code>, …\nReturns a mutable reference to the element at the …\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nReturns whether the vector has no elements.\nReturns whether the vector is at maximum capacity.\nReturns an iterator over immutable references to the …\nReturns an iterator over mutable references to the …\nReturns a reference to the last element in the vector, or …\nReturns a mutable reference to the last element in the …\nReturns the number of elements the vector currently …\nCreates a new empty <code>Vec</code> with maximum <code>CAPACITY</code> elements of …\nCreates immutable iterator.\nCreates mutable iterator.\nReturns (and removes) the last element from the vector, or …\nReturns (and removes) the last element from the vector if …\nAdds the given <code>value</code> to the end of the vector.\nResizes the vector to the <code>new_length</code>.")