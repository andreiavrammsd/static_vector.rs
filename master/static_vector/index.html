<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`static_vector`"><title>static_vector - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="static_vector" data-themes="" data-resource-suffix="" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../static.files/storage-82c7156e.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../static_vector/index.html">static_<wbr>vector</a><span class="version">0.3.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#static_vector" title="`static_vector`"><code>static_vector</code></a><ul><li><a href="#features" title="Features">Features</a></li><li><a href="#requirements" title="Requirements">Requirements</a></li><li><a href="#complexity" title="Complexity">Complexity</a></li><li><a href="#add-to-project" title="Add to project">Add to project</a></li><li><a href="#example" title="Example">Example</a></li><li><a href="#development-on-linux" title="Development on Linux">Development on Linux</a></li></ul></li></ul><h3><a href="#structs">Crate Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>static_vector</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/static_vector/lib.rs.html#1-1542">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="static_vector"><a class="doc-anchor" href="#static_vector">§</a><code>static_vector</code></h2>
<p><a href="https://github.com/andreiavrammsd/static_vector.rs/actions/workflows/ci.yml"><img src="https://github.com/andreiavrammsd/static_vector.rs/workflows/CI/badge.svg" alt="build" /></a>
<a href="https://codecov.io/gh/andreiavrammsd/static_vector.rs"><img src="https://codecov.io/gh/andreiavrammsd/static_vector.rs/graph/badge.svg?token=pCcpya0mZC" alt="codecov" /></a>
<a href="https://github.com/andreiavrammsd/static_vector.rs/actions/workflows/fuzz.yml"><img src="https://github.com/andreiavrammsd/static_vector.rs/workflows/FUZZ/badge.svg" alt="fuzz" /></a>
<a href="https://andreiavrammsd.github.io/static_vector.rs/"><img src="https://github.com/andreiavrammsd/static_vector.rs/workflows/DOC/badge.svg" alt="documentation" /></a>
<a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/License-MIT-blue.svg" alt="license" /></a></p>
<p>A no-std, stack-allocated vector with fixed capacity and dynamic length: <code>static_vector::Vec::&lt;T, CAPACITY&gt;</code>.</p>
<p><a href="struct.Vec.html" title="struct static_vector::Vec"><code>Vec</code></a> stores elements on the stack using a fixed-size array without heap allocations.</p>
<p>Aims to be suitable for low-level projects and to have an API as safe and explicit as possible.
The goal is to allocate only when needed. When first constructed, the vector will not allocate.</p>
<blockquote>
<p><strong>Note:</strong> It’s a learning project, so there are no guarantees.</p>
</blockquote>
<h3 id="features"><a class="doc-anchor" href="#features">§</a>Features</h3>
<ul>
<li>No heap allocation (<code>#![no_std]</code> compatible)</li>
<li>Supports iteration, mutable access, clearing, resizing</li>
<li>Compile-time enforced capacity</li>
</ul>
<h3 id="requirements"><a class="doc-anchor" href="#requirements">§</a>Requirements</h3>
<ul>
<li><code>CAPACITY</code> &gt; 0, otherwise <a href="struct.Vec.html#method.new" title="associated function static_vector::Vec::new"><code>Vec::new()</code></a> panics</li>
<li><code>T: Default</code> only if <a href="struct.Vec.html#method.set_len" title="method static_vector::Vec::set_len"><code>Vec::set_len()</code></a> is used</li>
</ul>
<h3 id="complexity"><a class="doc-anchor" href="#complexity">§</a>Complexity</h3>
<p>All operations are O(1) except:</p>
<div><table><thead><tr><th>Method</th><th>Time Complexity</th><th>Space Complexity</th></tr></thead><tbody>
<tr><td><code>clear</code></td><td>O(current length)</td><td>O(1)</td></tr>
<tr><td><code>set_len</code></td><td>O(new length - current length)</td><td>O(new length - current length)</td></tr>
<tr><td><code>extend_from_slice</code></td><td>O(slice length)</td><td>O(slice length)</td></tr>
<tr><td><code>append</code></td><td>O(other vector length)</td><td>O(other vector length)</td></tr>
</tbody></table>
</div><h3 id="add-to-project"><a class="doc-anchor" href="#add-to-project">§</a>Add to project</h3><div class="example-wrap"><pre class="language-bash"><code>cargo add static_vector --git https://github.com/andreiavrammsd/static_vector.rs</code></pre></div>
<p>This crate is not published on <a href="https://crates.io/">crates.io</a>.</p>
<h3 id="example"><a class="doc-anchor" href="#example">§</a>Example</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>static_vector::Vec;

<span class="kw">let </span><span class="kw-2">mut </span>vec = Vec::&lt;i32, <span class="number">3</span>&gt;::new();

vec.push(<span class="number">4</span>).unwrap();
vec.push(<span class="number">5</span>).unwrap();
vec.push(<span class="number">6</span>).unwrap();
<span class="macro">assert_eq!</span>(vec.len(), <span class="number">3</span>);
<span class="macro">assert_eq!</span>(vec.first(), <span class="prelude-val">Some</span>(<span class="kw-2">&amp;</span><span class="number">4</span>));

<span class="kw">let </span>sum_of_even_numbers = vec.iter().filter(|n| <span class="kw-2">*</span>n % <span class="number">2 </span>== <span class="number">0</span>).sum::&lt;i32&gt;();
<span class="macro">assert_eq!</span>(sum_of_even_numbers, <span class="number">10</span>);

vec.push(<span class="number">2</span>).unwrap_err();
<span class="macro">assert_eq!</span>(vec.len(), <span class="number">3</span>);

<span class="kw">match </span>vec.set_len(<span class="number">1</span>) {
    <span class="prelude-val">Ok</span>(()) =&gt; <span class="macro">assert_eq!</span>(vec.len(), <span class="number">1</span>),
    <span class="prelude-val">Err</span>(err) =&gt; <span class="macro">eprintln!</span>(<span class="string">"{:?}"</span>, err),
}

vec.clear();
<span class="macro">assert!</span>(vec.is_empty());</code></pre></div>
<p>See more examples in the documentation of <a href="struct.Vec.html" title="struct static_vector::Vec"><code>Vec</code></a>.</p>
<h3 id="development-on-linux"><a class="doc-anchor" href="#development-on-linux">§</a>Development on Linux</h3>
<ul>
<li>Install <a href="https://www.rust-lang.org/tools/install">Rust</a> (stable is used for main code, nightly is used only for code formatting and fuzz tests).</li>
<li>If using VS Code, see <a href="https://github.com/andreiavrammsd/static_vector.rs/tree/master/.vscode">setup</a>.</li>
<li>See <a href="https://github.com/andreiavrammsd/static_vector.rs/blob/master/Makefile">Makefile</a>.</li>
</ul>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.CapacityError.html" title="struct static_vector::CapacityError">Capacity<wbr>Error</a></dt><dd>Error for when the vector is full or the requested operation would need more space than the capacity.</dd><dt><a class="struct" href="struct.Iter.html" title="struct static_vector::Iter">Iter</a></dt><dd>Immutable iterator over a <a href="struct.Vec.html" title="struct static_vector::Vec"><code>Vec</code></a>.</dd><dt><a class="struct" href="struct.IterMut.html" title="struct static_vector::IterMut">IterMut</a></dt><dd>Mutable iterator over a <a href="struct.Vec.html" title="struct static_vector::Vec"><code>Vec</code></a>.</dd><dt><a class="struct" href="struct.Vec.html" title="struct static_vector::Vec">Vec</a></dt><dd>A stack-allocated vector with fixed capacity and dynamic length.</dd></dl></section></div></main></body></html>