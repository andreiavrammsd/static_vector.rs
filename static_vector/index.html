<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Static Vector"><title>static_vector - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="static_vector" data-themes="" data-resource-suffix="" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../static.files/storage-82c7156e.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../static_vector/index.html">static_<wbr>vector</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#static-vector" title="Static Vector">Static Vector</a><ul><li><a href="#features" title="Features">Features</a></li><li><a href="#requirements" title="Requirements">Requirements</a></li><li><a href="#complexity" title="Complexity">Complexity</a></li><li><a href="#example" title="Example">Example</a></li><li><a href="#development-on-linux" title="Development on Linux">Development on Linux</a></li></ul></li></ul><h3><a href="#structs">Crate Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>static_vector</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/static_vector/lib.rs.html#1-635">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="static-vector"><a class="doc-anchor" href="#static-vector">§</a>Static Vector</h2>
<p><a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/License-MIT-blue.svg" alt="license" /></a>
<a href="https://github.com/andreiavrammsd/static_vector.rs/actions/workflows/ci.yml"><img src="https://github.com/andreiavrammsd/static_vector.rs/workflows/CI/badge.svg" alt="build" /></a>
<a href="https://codecov.io/gh/andreiavrammsd/static_vector.rs"><img src="https://codecov.io/gh/andreiavrammsd/static_vector.rs/graph/badge.svg?token=pCcpya0mZC" alt="codecov" /></a>
<a href="https://andreiavrammsd.github.io/static_vector.rs/"><img src="https://img.shields.io/badge/Documentation-static_vector-4EC820.svg" alt="documentation" /></a></p>
<p>A no-std, stack-allocated vector with fixed capacity and dynamic length.</p>
<p><a href="struct.Vec.html" title="struct static_vector::Vec"><code>Vec</code></a> stores elements on the stack using a fixed-size array without heap allocations.</p>
<p>Aims to be suitable for low-level projects and to have an API as safe and explicit as possible.
The goal is to allocate only when needed. When first constructed, the vector will not allocate.</p>
<blockquote>
<p><strong>Note:</strong> It’s a learning project, so there are no guarantees.</p>
</blockquote>
<h3 id="features"><a class="doc-anchor" href="#features">§</a>Features</h3>
<ul>
<li>No heap allocation (<code>#![no_std]</code> compatible)</li>
<li>Supports iteration, mutable access, clearing, resizing</li>
<li>Compile-time enforced capacity</li>
</ul>
<h3 id="requirements"><a class="doc-anchor" href="#requirements">§</a>Requirements</h3>
<ul>
<li><code>T: Clone</code> for insertion: <a href="struct.Vec.html#method.push" title="method static_vector::Vec::push"><code>Vec::push()</code></a></li>
<li><code>T: Default</code> only if <a href="struct.Vec.html#method.set_len" title="method static_vector::Vec::set_len"><code>Vec::set_len()</code></a> is used</li>
</ul>
<h3 id="complexity"><a class="doc-anchor" href="#complexity">§</a>Complexity</h3>
<p>All operations are O(1) except:</p>
<div><table><thead><tr><th>Method</th><th>Time Complexity</th><th>Space Complexity</th></tr></thead><tbody>
<tr><td><code>clear</code></td><td>O(current length)</td><td>O(1)</td></tr>
<tr><td><code>set_len</code></td><td>O(new length - current length)</td><td>O(new length - current length)</td></tr>
</tbody></table>
</div><h3 id="example"><a class="doc-anchor" href="#example">§</a>Example</h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>static_vector::Vec;

<span class="kw">let </span><span class="kw-2">mut </span>vec = Vec::&lt;i32, <span class="number">3</span>&gt;::new();

vec.push(<span class="kw-2">&amp;</span><span class="number">4</span>).unwrap();
vec.push(<span class="kw-2">&amp;</span><span class="number">5</span>).unwrap();
vec.push(<span class="kw-2">&amp;</span><span class="number">6</span>).unwrap();
<span class="macro">assert_eq!</span>(vec.len(), <span class="number">3</span>);
<span class="macro">assert_eq!</span>(vec.first(), <span class="prelude-val">Some</span>(<span class="kw-2">&amp;</span><span class="number">4</span>));

<span class="kw">let </span>sum_of_even_numbers = vec.iter().filter(|n| <span class="kw-2">*</span>n % <span class="number">2 </span>== <span class="number">0</span>).sum::&lt;i32&gt;();
<span class="macro">assert_eq!</span>(sum_of_even_numbers, <span class="number">10</span>);

vec.push(<span class="kw-2">&amp;</span><span class="number">2</span>).unwrap_err();
<span class="macro">assert_eq!</span>(vec.len(), <span class="number">3</span>);

<span class="kw">match </span>vec.set_len(<span class="number">1</span>) {
    <span class="prelude-val">Ok</span>(()) =&gt; <span class="macro">assert_eq!</span>(vec.len(), <span class="number">1</span>),
    <span class="prelude-val">Err</span>(err) =&gt; <span class="macro">eprintln!</span>(<span class="string">"{:?}"</span>, err),
}

vec.clear();
<span class="macro">assert!</span>(vec.is_empty());</code></pre></div>
<h3 id="development-on-linux"><a class="doc-anchor" href="#development-on-linux">§</a>Development on Linux</h3>
<p>See <a href="https://github.com/andreiavrammsd/static_vector.rs/blob/master/Makefile">Makefile</a>.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.CapacityExceededError.html" title="struct static_vector::CapacityExceededError">Capacity<wbr>Exceeded<wbr>Error</a></dt><dd>Attempted to push to a full vector</dd><dt><a class="struct" href="struct.Iter.html" title="struct static_vector::Iter">Iter</a></dt><dd>Immutable iterator over a <a href="struct.Vec.html" title="struct static_vector::Vec"><code>Vec</code></a>.</dd><dt><a class="struct" href="struct.IterMut.html" title="struct static_vector::IterMut">IterMut</a></dt><dd>Mutable iterator over a <a href="struct.Vec.html" title="struct static_vector::Vec"><code>Vec</code></a>.</dd><dt><a class="struct" href="struct.LengthTooLargeError.html" title="struct static_vector::LengthTooLargeError">Length<wbr>TooLarge<wbr>Error</a></dt><dd>Attempted to resize the vector to a length greater than its fixed capacity.</dd><dt><a class="struct" href="struct.Vec.html" title="struct static_vector::Vec">Vec</a></dt><dd>A stack-allocated vector with fixed capacity and dynamic length.</dd></dl></section></div></main></body></html>