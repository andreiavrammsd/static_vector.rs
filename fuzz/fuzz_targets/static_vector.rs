#![no_main]
#![deny(warnings)]

use libfuzzer_sys::fuzz_target;
use static_vector::Vec;

fuzz_target!(|data: &[u8]| {
    let mut vec: Vec<u8, 125> = Vec::new();

    assert_eq!(vec.capacity(), 125);
    assert_eq!(vec.len(), 0);
    assert!(vec.is_empty());
    assert!(!vec.is_full());
    assert!(vec.first().is_none());
    assert!(vec.last().is_none());
    assert!(vec.pop().is_none());
    assert!(vec.pop_if(|_| true).is_none());

    let mut first_byte = None;
    let mut prev_byte = None;

    for (i, &byte) in data.iter().enumerate() {
        let is_full_before_push = vec.is_full();
        let result = vec.push(&byte);
        let is_full_after_push = vec.is_full();

        if first_byte.is_none() {
            first_byte = Some(byte);
        }
        assert_eq!(vec.first().unwrap(), &first_byte.unwrap());

        if is_full_before_push {
            assert_eq!(vec.len(), vec.capacity());
            assert!(result.is_err());
            assert!(!vec.is_empty());
            assert!(vec.is_full());
            assert_eq!(vec.last().unwrap(), &prev_byte.unwrap());
            assert!(vec.get(i).is_none());
            assert!(vec.get_mut(i).is_none());
        } else {
            assert_eq!(vec.len(), i + 1);
            assert!(result.is_ok());
            assert!(!vec.is_empty());

            if is_full_after_push {
                assert!(vec.is_full());
            } else {
                assert!(!vec.is_full());
            }

            assert_eq!(vec.last().unwrap(), &byte);
            assert_eq!(vec.get(i).unwrap(), &byte);
            assert_eq!(vec.get_mut(i).unwrap(), &byte);

            prev_byte = Some(byte);
        }
    }

    vec.as_mut_slice().fill_with(|| 1);
    assert_eq!(vec.as_slice().iter().sum::<u8>(), vec.len() as u8);

    vec.clear();
    for &byte in data {
        assert!(vec.push(&byte).is_ok());
        assert_eq!(vec.len(), 1);
        assert_eq!(vec.pop().unwrap(), byte);
        assert!(vec.is_empty());
    }

    vec.clear();
    for &byte in data {
        assert!(vec.push(&byte).is_ok());
        assert_eq!(vec.len(), 1);
        assert_eq!(vec.pop_if(|_b| true).unwrap(), byte);
        assert!(vec.is_empty());
    }
});
