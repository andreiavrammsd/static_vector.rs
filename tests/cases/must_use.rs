#![deny(warnings)]

use static_vector::Vec;

fn main() {
    let mut vec = Vec::<i32, 1>::new();
    vec.is_empty();
    vec.is_full();
    vec.capacity();
    vec.len();
    vec.push(&1);
    vec.set_len(1);
    vec.first();
    vec.first_mut();
    vec.last();
    vec.last_mut();
    vec.get(0);
    vec.get_mut(0);
    vec.pop();
    vec.pop_if(|&_i32| true);
    vec.iter();
    vec.iter_mut();

    Vec::<i32, 1>::new();
}
