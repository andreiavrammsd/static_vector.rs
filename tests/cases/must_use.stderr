error: unused return value of `static_vector::Vec::<T, CAPACITY>::is_empty` that must be used
 --> tests/cases/must_use.rs:7:5
  |
7 |     vec.is_empty();
  |     ^^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> tests/cases/must_use.rs:1:9
  |
1 | #![deny(warnings)]
  |         ^^^^^^^^
  = note: `#[deny(unused_must_use)]` implied by `#[deny(warnings)]`
help: use `let _ = ...` to ignore the resulting value
  |
7 |     let _ = vec.is_empty();
  |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::is_full` that must be used
 --> tests/cases/must_use.rs:8:5
  |
8 |     vec.is_full();
  |     ^^^^^^^^^^^^^
  |
help: use `let _ = ...` to ignore the resulting value
  |
8 |     let _ = vec.is_full();
  |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::len` that must be used
 --> tests/cases/must_use.rs:9:5
  |
9 |     vec.len();
  |     ^^^^^^^^^
  |
help: use `let _ = ...` to ignore the resulting value
  |
9 |     let _ = vec.len();
  |     +++++++

error: unused `Result` that must be used
  --> tests/cases/must_use.rs:10:5
   |
10 |     vec.push(&1);
   |     ^^^^^^^^^^^^
   |
   = note: this `Result` may be an `Err` variant, which should be handled
help: use `let _ = ...` to ignore the resulting value
   |
10 |     let _ = vec.push(&1);
   |     +++++++

error: unused `Result` that must be used
  --> tests/cases/must_use.rs:11:5
   |
11 |     vec.set_len(1);
   |     ^^^^^^^^^^^^^^
   |
   = note: this `Result` may be an `Err` variant, which should be handled
help: use `let _ = ...` to ignore the resulting value
   |
11 |     let _ = vec.set_len(1);
   |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::first` that must be used
  --> tests/cases/must_use.rs:12:5
   |
12 |     vec.first();
   |     ^^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
12 |     let _ = vec.first();
   |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::last` that must be used
  --> tests/cases/must_use.rs:13:5
   |
13 |     vec.last();
   |     ^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
13 |     let _ = vec.last();
   |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::get` that must be used
  --> tests/cases/must_use.rs:14:5
   |
14 |     vec.get(0);
   |     ^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
14 |     let _ = vec.get(0);
   |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::get_mut` that must be used
  --> tests/cases/must_use.rs:15:5
   |
15 |     vec.get_mut(0);
   |     ^^^^^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
15 |     let _ = vec.get_mut(0);
   |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::pop` that must be used
  --> tests/cases/must_use.rs:16:5
   |
16 |     vec.pop();
   |     ^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
16 |     let _ = vec.pop();
   |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::pop_if` that must be used
  --> tests/cases/must_use.rs:17:5
   |
17 |     vec.pop_if(|&_i32| true);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
17 |     let _ = vec.pop_if(|&_i32| true);
   |     +++++++

error: unused `static_vector::Iter` that must be used
  --> tests/cases/must_use.rs:18:5
   |
18 |     vec.iter();
   |     ^^^^^^^^^^
   |
   = note: must consume iterator
help: use `let _ = ...` to ignore the resulting value
   |
18 |     let _ = vec.iter();
   |     +++++++

error: unused `static_vector::IterMut` that must be used
  --> tests/cases/must_use.rs:19:5
   |
19 |     vec.iter_mut();
   |     ^^^^^^^^^^^^^^
   |
   = note: must consume iterator
help: use `let _ = ...` to ignore the resulting value
   |
19 |     let _ = vec.iter_mut();
   |     +++++++

error: unused return value of `static_vector::Vec::<T, CAPACITY>::new` that must be used
  --> tests/cases/must_use.rs:21:5
   |
21 |     Vec::<i32, 1>::new();
   |     ^^^^^^^^^^^^^^^^^^^^
   |
help: use `let _ = ...` to ignore the resulting value
   |
21 |     let _ = Vec::<i32, 1>::new();
   |     +++++++
